cpp/serve/engine_actions/batch_verify.cc

```diff
@@ -0,0 +1,417 @@
+          draft_mstate->draft_output_tokens, draft_mstate->draft_output_token_prob,
+          draft_mstate->draft_output_prob_dist);
+      estate->stats.total_accepted_length += accept_length;
+      // minus one because the last draft token has no kv cache entry
+      int rollback_length = verify_end - verify_start - accept_length - 1;
+      // in case of all accepted
+      rollback_length = std::max(0, rollback_length);
+      // rollback kv cache
+      if (rollback_length > 0) {
+        models_[verify_model_id]->PopNFromKVCache(
+            rstates[i - 1]->mstates[verify_model_id]->internal_id, rollback_length);
+        models_[draft_model_id]->PopNFromKVCache(
+            rstates[i - 1]->mstates[draft_model_id]->internal_id, rollback_length);
+      }
+    }
+
+    // clear the draft model states
+    for (int i = 0; i < num_requests; ++i) {
+      rstates[i]->mstates[draft_model_id]->draft_output_tokens.clear();
+      rstates[i]->mstates[draft_model_id]->draft_output_token_prob.clear();
+      rstates[i]->mstates[draft_model_id]->draft_output_prob_dist.clear();
+    }
+
+    auto tend = std::chrono::high_resolution_clock::now();
+    estate->stats.engine_total_decode_time += static_cast<double>((tend - tstart).count()) / 1e9;
+
+    return requests;
+  }
+
+ private:
+  bool CanDecode(int num_requests) {
+    int num_available_pages = models_[0]->GetNumAvailablePages();
+    return num_requests <= num_available_pages;
+  }
+
+  /*! \brief Check if the drafts can be verified under conditions. */
+  bool CanVerify(EngineState estate, int num_verify_req, int total_draft_length,
+                 int num_required_pages, int num_available_pages) {
+    int num_running_requests = estate->running_queue.size();
+    ICHECK_LE(num_running_requests, kv_cache_config_->max_num_sequence);
+
+    // No exceeding of the maximum allowed requests that can
+    // run simultaneously.
+    if (num_running_requests + num_verify_req > kv_cache_config_->max_num_sequence) {
+      return false;
+    }
+
+    // NOTE: The conditions are heuristic and can be revised.
+    // Cond 1: total input length <= max allowed single sequence length.
+    // Cond 2: at least one verify can be performed.
+    // Cond 3: number of total tokens does not exceed the limit
+    int new_batch_size = num_running_requests + num_verify_req;
+    return total_draft_length <= max_single_sequence_length_ &&
+           num_required_pages <= num_available_pages &&
+           estate->stats.current_total_seq_len + total_draft_length <=
+               kv_cache_config_->max_total_sequence_length;
+  }
+
+  /*!
+   * \brief Decide whether to run verify for the draft of each request.
+   * \param estate The engine state.
+   * \return The drafts to verify, together with their respective
+   * state and input length.
+   */
+  std::tuple<Array<Request>, Array<RequestState>, std::vector<int>, int> GetDraftsToVerify(
+      EngineState estate) {
+    const int verify_model_id = 0;
```

Comment:
Can we make`verify_model_id(s)` and `draft_model_id` part of class members? I see repetitive definition in several functions, which might be hard to maintain if we support multiple SSMs in the future.

Reply:
Yes I agree that we can lift them as const class members.